components/script/dom/dommatrixreadonly.rs

```diff
@@ -0,0 +1,687 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use dom::bindings::cell::DOMRefCell;
+use dom::bindings::codegen::Bindings::DOMMatrixBinding::{DOMMatrixInit, DOMMatrixMethods};
+use dom::bindings::codegen::Bindings::DOMMatrixReadOnlyBinding::{DOMMatrixReadOnlyMethods, Wrap};
+use dom::bindings::codegen::Bindings::DOMPointBinding::DOMPointInit;
+use dom::bindings::error;
+use dom::bindings::error::Fallible;
+use dom::bindings::global::GlobalRef;
+use dom::bindings::js::Root;
+use dom::bindings::reflector::{reflect_dom_object, Reflectable, Reflector};
+use dom::dommatrix::DOMMatrix;
+use dom::dompoint::DOMPoint;
+use euclid::Matrix4D;
+use euclid::Point4D;
+use std::cell::Cell;
+use std::f64;
+
+#[dom_struct]
+pub struct DOMMatrixReadOnly {
+    reflector_: Reflector,
+    matrix: DOMRefCell<Matrix4D<f64>>,
+    is2D: Cell<bool>,
+}
+
+impl DOMMatrixReadOnly {
+    #[allow(unrooted_must_root)]
+    pub fn new(global: GlobalRef, is2D: bool, matrix: Matrix4D<f64>) -> Root<Self> {
+        let dommatrix = Self::new_inherited(is2D, matrix);
+        reflect_dom_object(box dommatrix, global, Wrap)
+    }
+
+    pub fn new_inherited(is2D: bool, matrix: Matrix4D<f64>) -> Self {
+        DOMMatrixReadOnly {
+            reflector_: Reflector::new(),
+            matrix: DOMRefCell::new(matrix),
+            is2D: Cell::new(is2D),
+         }
+    }
+
+    // https://drafts.fxtf.org/geometry-1/#dom-dommatrixreadonly-dommatrixreadonly
+    pub fn Constructor(global: GlobalRef) -> Fallible<Root<Self>> {
+        entries_to_matrix(&[1.0, 0.0, 0.0, 1.0, 0.0, 0.0])
+            .map(|(is2D, matrix)| {
+                Self::new(global, is2D, matrix)
+            })
+    }
+
+    // https://drafts.fxtf.org/geometry-1/#dom-dommatrixreadonly-dommatrixreadonly-numbersequence
+    pub fn Constructor_(global: GlobalRef, entries: Vec<f64>) -> Fallible<Root<Self>> {
+        entries_to_matrix(&entries[..])
+            .map(|(is2D, matrix)| {
+                Self::new(global, is2D, matrix)
+            })
+    }
+
+    // https://drafts.fxtf.org/geometry-1/#dom-dommatrixreadonly-frommatrix
+    pub fn FromMatrix(global: GlobalRef, other: &DOMMatrixInit) -> Fallible<Root<Self>> {
+        dommatrixinit_to_matrix(&other)
+            .map(|(is2D, matrix)| {
+                Self::new(global, is2D, matrix)
+            })
+    }
+
+    fn to_DOMMatrix(&self) -> Root<DOMMatrix> {
```

Comment:
> Another way would be to add public accessors to DOMMatrixReadOnly for matrix and is2D, so that a new one can be constructed from those parts inside from_readonly.

That is what I had in mind. Just add two new methods on `DOMMatrixReadOnly`, `matrix` and `is_2d`.